### PR TITLE
feat(examples-plugins): add example plugin benchmark-js

### PR DESCRIPTION
**Pull request by BioPhoton**
This PR adds `benchmark` to the example plugins (and later move it into official plugins).

> [!NOTE]
> This PR includes 2 libs for benchmark testing. At them moment I'm unsure about the results and opened https://github.com/tinylibs/tinybench/issues/71.
> I will also wait for a couple of runs in the CI and maybe run more tests on my local machine. In a follow up PR I want to move it under our released plugins.

This plugin should execute the already existing benchmarks and generate audits of them.

In this PR:
- [x] moved mock folder
- [x] add `benchmark-js` to `examples/plugins`
  - [x] tests
  - [x] [docs](https://github.com/code-pushup/cli/blob/c89163452ef9ac2cb8114d51f1527a46a741710e/examples/plugins/src/benchmark-js/README.md)
  - [x] logic
  - [x] add target `perf`
- [x] introduce tools to run perf benchmarks more organised
- [x] refactor `utils/perf` to use new tooling
- [x] added error if `--tsconfig` file could not resolve + tests

Example md report:
audits-readme-example

**Related:**  
- [#tinylibs/tinybench/issues/71](https://github.com/tinylibs/tinybench/issues/71)

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -170,6 +170,8 @@ jobs:
           cache: npm
       - name: Install dependencies
         run: npm ci
+      - name: Check git version
+        run: git -v
```
Comment (BioPhoton) [inline .github/workflows/ci.yml:174]:
```suggestion
```

File: packages/utils/src/lib/execute-process.ts [hunk 2]
```diff
@@ -138,6 +138,7 @@ export function executeProcess(cfg: ProcessConfig): Promise<ProcessResult> {
   const start = performance.now();
 
   return new Promise((resolve, reject) => {
+    // @TODO add logging `Spawn process ${command} with args ${args?.join(' ')}`
```
Comment (BioPhoton) [inline packages/utils/src/lib/execute-process.ts:141]:
```suggestion
```

File: examples/plugins/src/benchmark-js/src/benchmark.suite-runner.integration.test.ts [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+            () => new Promise(resolve => setTimeout(resolve, 5)),
+          ],
+          [
+            'slower-implementation',
+            () => new Promise(resolve => setTimeout(resolve, 30)),
+          ],
+        ],
+      }),
+    ).resolves.toStrictEqual(expect.arrayContaining([]));
+  }, 20_000);
```
Comment (matejchalk) [inline examples/plugins/src/benchmark-js/src/benchmark.suite-runner.integration.test.ts:22]: Does this really need 20 seconds to run? :raised_eyebrow: 

Same goes for: `tinybench.suite-runner.integration.test.ts`

File: code-pushup.config.ts [hunk 4]
```diff
@@ -30,6 +32,11 @@ const envSchema = z
   .partial();
 const env = await envSchema.parseAsync(process.env);
 
+const benchmarkJsSuitNames = [
```
[the remark this answers is not in the document]
Comment (BioPhoton) [reply code-pushup.config.ts:36]: replaced them

File: project.json [hunk 5]
```diff
@@ -75,6 +75,21 @@
...
+          "projects": [
+            "cli",
+            "plugin-eslint",
+            "plugin-coverage",
+            "examples-plugins",
+            "react-todos-app"
+          ],
+          "target": "build"
+        }
+      ]
```
[the remark this answers is not in the document]
Comment (BioPhoton) [reply project.json]: removed as in https://github.com/code-pushup/cli/pull/273#discussion_r1537713960

File: examples/plugins/src/benchmark-js/src/benchmark.suite-runner.integration.test.ts (hunk 3 above)
Comment (BioPhoton) [reply examples/plugins/src/benchmark-js/src/benchmark.suite-runner.integration.test.ts]: down to `15_000`

File: packages/utils/perf/crawl-file-system/index.ts [hunk 6]
```diff
@@ -1,103 +1,89 @@
-import * as Benchmark from 'benchmark';
+import chalk from 'chalk';
 import { join } from 'node:path';
-import {
-  CrawlFileSystemOptions,
-  crawlFileSystem,
-} from '../../src/lib/file-system';
+import yargs from 'yargs';
+// eslint-disable-next-line @nx/enforce-module-boundaries
+import { type CrawlFileSystemOptions } from '../../../../dist/packages/utils';
```
[the remark this answers is not in the document]
Comment (BioPhoton) [reply packages/utils/perf/crawl-file-system/index.ts:5]: This is a mistake. Thanks for pointing it out.

File: packages/utils/perf/score-report/index.ts [hunk 7]
```diff
@@ -1,201 +1,62 @@
-import * as Benchmark from 'benchmark';
-import { Report } from '@code-pushup/models';
-import { scoreReport } from '../../src/lib/reports/scoring';
+import yargs from 'yargs';
+// eslint-disable-next-line @nx/enforce-module-boundaries
+import { scoreReport } from '../../../../dist/packages/utils';
```
[the remark this answers is not in the document]
Comment (BioPhoton) [reply packages/utils/perf/score-report/index.ts]: This is a mistake. Thanks for pointing it out.

File: examples/plugins/src/js-benchmarking/README.md [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+const suiteConfig = {
+  suiteName: 'glob',
+  targetImplementation: 'version-2',
+  cases: [
+    ['version-1', () => new Promise(resolve => setTimeout(resolve, 30))],
+    ['version-2', () => new Promise(resolve => setTimeout(resolve, 10))],
+    ['version-3', () => new Promise(resolve => setTimeout(resolve, 20))],
+  ],
+};
+```
```
Comment (matejchalk) [inline examples/plugins/src/js-benchmarking/README.md:54]: In Markdown, you have to use indentation to nest lists:

```suggestion
   1. Create benchmark suite:

      ```ts
      // typescript
      const suiteConfig = {
        suiteName: 'glob',
        targetImplementation: 'version-2',
        cases: [
          ['version-1', () => new Promise(resolve => setTimeout(resolve, 30))],
          ['version-2', () => new Promise(resolve => setTimeout(resolve, 10))],
          ['version-3', () => new Promise(resolve => setTimeout(resolve, 20))],
        ],
      };
      ```
```

### Before

![image](https://github.com/code-pushup/cli/assets/34691111/9d4b693e-e4ac-4326-9dd6-551895fc28ea)


### After

![image](https://github.com/code-pushup/cli/assets/34691111/2b9dc01a-907e-49d4-a4b7-d29fed3d35ea)

File: package.json [hunk 9]
```diff
@@ -50,6 +50,7 @@
     "@isaacs/cliui": "^8.0.2",
     "@poppinss/cliui": "^6.4.0",
     "@swc/helpers": "0.5.3",
+    "benny": "^3.7.1",
```
Comment (matejchalk) [inline package.json:53]: Should be a dev dependency since it's only used in example plugins.

**Review by matejchalk — APPROVED** [review]

File: examples/plugins/src/js-benchmarking/README.md (hunk 8 above)
Comment (BioPhoton) [reply examples/plugins/src/js-benchmarking/README.md]: Thanks for the detailed explanation. TBH I never got the second intention. :D

**Pull request closed without merging**